### PR TITLE
Add white-list group filter to wavefield writer.

### PR DESCRIPTION
**Pull request by krenzland**
See documentation:

 Similar to the previous parameter, OutputGroups can be used to whitelist a set of
mesh groups (as specified in the xdmf mesh file) that are included in the wavefieldoutput.
Cells whose group is not mentioned are not included in the output.
This feature works with OutputRegionBounds, only cells that satisfy both criteria are included.
It looks like this:

`   OutputGroups = 1 2 ! only include groups 1 and 2`

Note that this is a very old commit from 2020 that I just found lying around - not sure why I never merged this. I just cherry picked this from an old branch and merged with master. The feature was tested then; I haven't run extensive tests now but a small scale test works. The code is very simple as well and just piggy-backs on the old region filter, so I see no reason why it would work on more complicated setups as well.

### CODE REVIEW

File: Documentation/wave-field-output.rst [hunk 1]
```diff
@@ -57,14 +57,27 @@ OutputRegionBounds
...
 
 .. code-block:: Fortran
 
    OutputRegionBounds = xMin xMax yMin yMax zMin zMax
 
+OutputGroups
+------------------
+
+Similar to the previous parameter, OutputGroups can be used to whitelist a set of
+mesh groups (as specified in the xdmf mesh file) that are included in the wavefieldoutput.
```
Comment (Thomas-Ulrich) [inline Documentation/wave-field-output.rst:72]: wavefield output

File: src/Numerical_aux/typesdef.f90 [hunk 2]
```diff
@@ -1091,6 +1091,7 @@ MODULE TypesDef
                                                                                 !< .FALSE. = do not integrate and output for this variable
      REAL                         ,POINTER  :: OutputRegionBounds(:)            !< Region for which the output should be written
                                                                                 !< Format is xMin, xMax, yMin, yMax, zMin, zMax
+     INTEGER, ALLOCATABLE :: OutputGroups(:) !< Stores an array of group ids that should be included in output. All others are excluded.
```
Comment (Thomas-Ulrich) [inline src/Numerical_aux/typesdef.f90:1094]: in the volume output.

File: src/Reader/readpar.f90 [hunk 3]
```diff
@@ -2772,7 +2775,9 @@ SUBROUTINE readpar_output(EQN,DISC,IO,MPI,CalledFromStructCode)
           ENDIF
       END IF
 
-      ALLOCATE(IO%IntegrationMask(9),STAT=allocstat )                        !
+    IO%OutputGroups = pack(OutputGroups, OutputGroups >= 0)
+
+	  ALLOCATE(IO%IntegrationMask(9),STAT=allocstat )                        !
```
Comment (Thomas-Ulrich) [inline src/Reader/readpar.f90:2780]: fix code alignment

Comment (Thomas-Ulrich) [reply src/Reader/readpar.f90:2780]: Also don't you need to allocate IO%OutputGroups first?

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 4]
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
 
 		subVertices.resize(oldToNewVertexMap.size());
 
-		// Loop over the map and assign the vertices
-		for (std::map<int,int>::iterator it=oldToNewVertexMap.begin(); it!=oldToNewVertexMap.end(); ++it)
-			subVertices.at(it->second) = &allVertices[it->first];
+		for (auto & it : oldToNewVertexMap) {
+            subVertices.at(it.second) = &allVertices[it.first];
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:264]: fix alignment? (tabulations...)

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 5]
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
 
 		// m_map will store a new map from new cell index to dof index
 		// the old map is contained in the "map" variable - which is a map from old
 		//    cell index to dof index
 		m_map = new unsigned int[numElems];
 
-		// Extract elements based on the region specified
+		// Extract elements based on the specified group or region
 		for (size_t i = 0; i < numElems; i++) {
+		    // Note: For some reason, an alternative name for groupId is material.
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:235]: should we change that?

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 6]
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
-	m_extractRegion = outputRegionBounds[0] != 0.0 ||
-		outputRegionBounds[1] != 0.0 || outputRegionBounds[2] != 0.0 ||
-		outputRegionBounds[3] != 0.0 || outputRegionBounds[4] != 0.0 ||
-		outputRegionBounds[5] != 0.0;
-
-	if (m_extractRegion) {
-		/** Extract the elements and vertices based on user given bounds */
-		// Reference to the vector containing all the elements
-		const std::vector<Element>& allElements = meshReader.getElements();
+	const bool extractBox = outputRegionBounds[0] != 0.0
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:214]: extractBoxEnabled and extractGroupEnabled?

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 7]
```diff
@@ -202,38 +205,43 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
-				vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[1]].coords) ||
-				vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[2]].coords) ||
-				vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[3]].coords)) {
-
+			const bool isInRegion = !extractBox
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[0]].coords)
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[1]].coords)
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[2]].coords)
+			        || vertexInBox(outputRegionBounds, allVertices[allElements[i].vertices[3]].coords);
+			const bool isInGroup = !extractGroup || outputGroups.count(groupId) > 0; // TODO(Lukas) Implement this.
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:243]: why the comment?

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 8]
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
+		}
 
 		numElems = subElements.size();
 		numVerts = subVertices.size();
 
 		meshRefiner = new refinement::MeshRefiner<double>(subElements, subVertices,
 			oldToNewVertexMap, *tetRefiner);
-	} else {
-		meshRefiner = new refinement::MeshRefiner<double>(meshReader, *tetRefiner);
+    } else {
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:272]: fix alignment? (tabulations...)

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 9]
```diff
@@ -252,28 +260,32 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
+        meshRefiner = new refinement::MeshRefiner<double>(meshReader, *tetRefiner);
 		m_map = map;
 	}
 
+    if (numElems == 0) {
+        logError() << "All cells are excluded in the WaveFieldWriter.";
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:278]: WaveFieldWriter: All cells have been filtered out (OutputRegionBounds and OutputGroups).

**Review by Thomas-Ulrich — CHANGES_REQUESTED** [review]

File: src/Reader/readpar.f90 (hunk 3 above)
Comment (krenzland) [reply src/Reader/readpar.f90:2780]: I don't think so - pack seems to allocate. All examples that I can find where pack is used on allocateable arrays do not use allocate manually.

Comment (Thomas-Ulrich) [reply src/Reader/readpar.f90:2780]: ok you're right (I was too lazy to google).

**Comment by Thomas-Ulrich** [comment]
Just did a quick test (with both extract options). It works:
image

**Comment by krenzland** [comment]
Thanks, Thomas!

File: src/ResultWriter/WaveFieldWriter.cpp [hunk 10]
```diff
@@ -208,21 +208,22 @@ void seissol::writer::WaveFieldWriter::init(unsigned int numVars,
...
-		const std::vector<Element>& allElements = meshReader.getElements();
+	const bool isExtractBoxEnabled = outputRegionBounds[0] != 0.0
+                                   || outputRegionBounds[1] != 0.0
+                                   || outputRegionBounds[2] != 0.0
+                                   || outputRegionBounds[3] != 0.0
+                                   || outputRegionBounds[4] != 0.0
+                                   || outputRegionBounds[5] != 0.0;
+
+  // If at least one group is explicitly enabled, extract
+  const bool isExtractGroupEnabled = !outputGroups.empty();
```
Comment (Thomas-Ulrich) [inline src/ResultWriter/WaveFieldWriter.cpp:219]: alignment not fixed (check in GitHub).

File: src/Geometry/NetcdfReader.h [hunk 11]
```diff
@@ -258,7 +258,7 @@ class NetcdfReader : public MeshReader
 		ElemBoundaries* elemBoundaries = new ElemBoundaries[maxSize];
 		ElemNeighborRanks* elemNeighborRanks = new ElemNeighborRanks[maxSize];
 		ElemMPIIndices* elemMPIIndices = new ElemMPIIndices[maxSize];
-		ElemMaterial* elemMaterial = new ElemMaterial[maxSize];
+		ElemGroup* elemMaterial = new ElemGroup[maxSize];
```
Comment (Thomas-Ulrich) [inline src/Geometry/NetcdfReader.h:261]: Have you considered also changing also variable names (e.g. elemMaterial)?

File: src/Geometry/PUMLReader.cpp [hunk 12]
```diff
@@ -381,7 +381,7 @@ void seissol::PUMLReader::getMesh(const PUML::TETPUML &puml)
 			m_elements[i].mpiIndices[FACE_PUML2SEISSOL[j]] = 0;
 		}
 
-		m_elements[i].material = material[i];
+		m_elements[i].group = material[i];
```
Comment (Thomas-Ulrich) [inline src/Geometry/PUMLReader.cpp:384]: material --> group?

**Review by Thomas-Ulrich — COMMENTED** [review]
Thx, but alignments are not yet fixed...

**Comment by krenzland** [comment]
I've fixed the alignment. The entire file was broken because it randomly mixes tabs and spaces. I just replaced all tabs with spaces. Larger diff, but otherwise it's impossible to get the alignment working :( (It was broken before my PR)

**Comment by Thomas-Ulrich** [comment]
by the way, at this point you could clang-format the file (using format prescribed in dr-cpp branch).

**Pull request merged**